libwacom/libwacom-database.c

```diff
@@ -279,8 +328,81 @@ parse_stylus_id(const char *str, WacomStylusId *id)
 	return ret;
 }
 
+static char *
+string_or_fallback (GKeyFile *keyfile, const char *group,
+		    const char *key, const char *fallback, GError *error)
+{
+	GError *local_error = NULL;
+	bool ret;
+
+	ret = g_key_file_has_key(keyfile, group, key, &local_error);
+	if (local_error) {
+		g_warning ("String fallback error: %s\n", local_error->message);
+		g_clear_error (&local_error);
+	}
+	if (ret)
+		return g_key_file_get_string(keyfile, group, key, &error);
+
+	return g_strdup(fallback);
+}
+
+static int
+int_or_fallback (GKeyFile *keyfile, const char *group,
+		 const char *key, int fallback, GError *error)
+{
+	GError *local_error = NULL;
+	bool ret;
+
+	ret = g_key_file_has_key(keyfile, group, key, &local_error);
+	if (local_error) {
+		g_warning ("int fallback error: %s\n", local_error->message);
+		g_clear_error (&local_error);
+	}
+
+	if (ret)
+		return g_key_file_get_integer(keyfile, group, key, &error);
+
+	return fallback;
+}
+
+static bool
+boolean_or_fallback (GKeyFile *keyfile, const char *group,
+		     const char *key, bool fallback, WacomStylus *stylus, GError *error)
+{
+	GError *local_error = NULL;
+	bool ret;
+
+	ret = g_key_file_has_key(keyfile, group, key, &local_error);
+	if (local_error) {
+		g_warning ("boolean fallback error: %s\n", local_error->message);
+		g_clear_error (&local_error);
+	}
+
+	if (ret)
+		return g_key_file_get_boolean(keyfile, group, key, &error);
+
+	return fallback;
+}
+
+static gchar **
+stylus_ids_as_hex(GArray *array)
+{
+	WacomStylusId *id;
+
+	if (array == NULL || array->len == 0)
+		return NULL;
+
+	GArray *a = g_array_new(true, false, array->len);
+	for (guint i = 0; i < array->len; i++) {
+		id = &g_array_index(array, WacomStylusId, i);
+		g_array_append_vals(a, g_strdup_printf("0x%08x", id->tool_id), 1);
```

Comment:
see related comment below: best to append the vendor id here too to make this a full stylus id in the form `0x1234:0xabcd`

Reply:
The way this appeared, it is your last comment. So by below I think you're referring to the comment right above this, just double checking that I understand you right.

Reply:
I just noticed #871 - so you can ignore the comments I just made. Sorry about that I was going through my mail oldest to newest.